enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,891 @@
+* Modifying any of the existing deployments of Machine API within clusters
+* Deleting or modifying any existing resources within clusters (until a user chooses to do so)
+* Force creation of new Machine API machines when new Cluster API MachineSets are being created[^4]
+* Port any feature from Cluster API into Machine API
+
+[^4]: We want to promote, over time, a reduction in the number of Machine API Machines, as such,
+any resources created directly in Cluster API, will not be reflected into Machine API, unless they have an owner,
+and, that owner is already reflected into Machine API.
+
+## Proposal
+
+We will implement a two-way sync controller that synchronises Machines and related resources between the Machine API and Cluster API equivalents.
+
+Using a new field on the Machine API resource, users will be able to choose which API is authoritative.
+Any discrepancy between the non-authoritative resource and the authoritative resource will be overwritten by the sync controller to match the authoritative resource.
+
+When the API is non-authoritative, the controllers should be paused (ignore the resource),
+allowing the authoritative API’s controller to perform the reconciliation actions required.
+
+### Workflow Description
+
+In this workflow, the **cluster admin** (or equivalent controller, e.g. Hive) is responsible for the infrastructure provisioning for the cluster.
+Machine API and Cluster API represent different API group versions, but with similar resources.
+This workflow allows switching management of the infrastructure resources (eg EC2 instances) from the former API, to the new API.
+
+When the cluster admin wishes to migrate a MachineSet from Machine API to Cluster API, the following procedure is required:
+1. Identify the MachineSet to migrate to Cluster API
+1. Use the `oc edit` or a patch command to update the value of the `spec.authoritativeAPI` field to `ClusterAPI`
+1. The migration controller verifies that the `Snychonrized` condition is currently set to `True`, verifying no long standing synchronization errors
+1. The `status.authoritativeAPI` field is updated to `Migrating` by the migration controller
+1. The Machine API controller acknowledges the change and sets the `Paused` condition to `True`
+1. The sync controller ensures the latest changes are synchronised between the old authoritative resource and the new, the `status.synchronizedGeneration` is then updated to the current generation of the old authoritative resource
+1. The migration controller verifies that the move from Machine API to Cluster API is valid by checking that the synchronized generation is up to date
+1. The migration controller updates `status.authoritativeAPI`  to `ClusterAPI`
+1. The Cluster API controller takes over management of the MachineSet going forwards
+
+To migrate back from Cluster API to Machine API, the procedure is the same, however, the value of the field should be set to `MachineAPI`.
```

Comment:
From the goals section:

> Allow users to migrate Machines between Machine API and Cluster API to test the new API

However, upon re-reading it, the full motivation isn't spelled out.

The idea for going back is an escape hatch in case something is broken or missing in CAPI that does work/exists in MAPI.

If there was consensus around dropping that it would indeed save us some work, but I think right now we'd like to be super conservative in moving forward, since breakages could render clusters unusable.

Reply:
Yeah, we are trying to be very very conservative here so that we don't end up getting anyone stuck in a situation where they are migrated to Cluster API and can't move backwards if there's a bug. It would be preferable to not have to deal with the reverse situation, and we are of course prioritising the forward direction, but I would still be keen to see the reverse direction implemented unless we suddenly become super confident in our migration and that no one would need the reverse

Reply:
In my head I'm trying to relate this back to specific kinds of things that could go wrong and how the user could get out of them.
If we really messed and and ended up reprovisioning all the Machines that would be bad, but at that point switching back (and potentially reprovisioning them all again, heh) wouldn't be a solution.
Machines aren't supposed to do much once they're already provisioned, so as long as they can just sit there there's not a lot to go wrong.
If we're unable to create new Machines in a MachineSet for whatever reason then the user always has the escape hatch of creating a new MAPI MachineSet rather than reverse-migrating the existing one.
The only thing I can think of is if we're unable to delete a Machine using CAPI (and not because we've thrown away some data that we needed, since we likely can't recover that), then migrating it back to MAPI could be a workaround.

Reply:
> I think right now we'd like to be super conservative in moving forward

I can't help being reminded of what they used to call the [MIT approach](https://www.dreamsongs.com/RiseOfWorseIsBetter.html).

> They must scale down the old MachineSet and remove it, because there's no backwards migration, this is unnecessary disruption

I don't agree with this part; I'm not aware of any rule that you are only allowed to have one MachineSet. So the user need not do this step.

> So really, we are talking about reversing the spec only as the part of this to drop right?

Technically I only asked about forbidding a change of the `authoritativeAPI` field from CAPI to MAPI.
But you're right that this also calls into question the value of reverse-migrating the spec if nothing is going to use it.

Really though, I'd love to do so much less than that.